Task 86/orthographic camera

krupitskas opened the pull request.

Commit:
[#86] Projection placeholders, vulkan resizing fixed

Commit:
[#86] Still need documentation

Commit:
[#86] Updated API

Commit:
[#86] Documentation added, fixed perspective mouse bug

Commit:
[#86] Moved crossproduct calculation out

Commit:
[#86] Constructor reordering

Commit:
[#86] Constructor reordering

Commit:
[#86] Removed geometry folder, changed mesh creation, camera refactoring

Commit:
[#86] Clang fix

Commit:
[#86] Moved camera stuff to different folders, fixed low end gpu crash.

Commit:
[#86] Camera rotation added, comments and functions refactoring _review_

Commit:
[#86] Quaternion camera, isvisible for mesh and SQT for mesh

[#86] At least moving front and right / left. Need to fix Z rotation

[#86] Quaternion camera working but still need understand -Y. Also fixed inversed world

[#86] Still need understand z rotating problem

[#86] Added rotate around pos, upvector partially fix

[#86] More camera work, interface beautifying, setposition fix

[#86] Need correct set direction vector functionality

[#86] fixing Rotate between vectors

Commit:
[#86] Correct rotations done

Commit:
[#86] Transform class for multiple transform entities

[#86] Linux fix

[#86] Debugging rot, squash it

[#86] Correct rotations, still need fix mouse

[#86] Investigating fps camera error

[#86] Fixed rotations

Commit:
[#86] Inversed tc for sphere and box, more documentation, remove conjugate quaternion

Commit:
[#86] comments and ortho fix

Commit:
[#86] Selfreview little improvements

ilia-glushchenko commented on UnicornRender/utility/include/unicorn/utility/Math.hpp:
```diff
@@ -0,0 +1,51 @@
+ * @param[in]  worldX  world X axis
+ * @param[in]  worldZ  world Z axis
+ *
+ * Realization : realtime rendering 3rd edition, 79 page
```
'Original paper from Real-time rendering 3-rd edition, page 79' mm?

MrDmitry replied:
yeah, _realization_ is just the wrong word. Also add author information

ilia-glushchenko replied:
Implementation!!!

ilia-glushchenko commented on UnicornRender/utility/include/unicorn/utility/Math.hpp:
```diff
@@ -0,0 +1,51 @@
+ *
+ * @param[in] direction new direction
+ * @param[in] desiredUp new up vector
+ *
```
worldY, worldX?

ilia-glushchenko commented on UnicornRender/utility/sources/Math.cpp:
```diff
@@ -0,0 +1,73 @@
+    start = normalize(start);
+    dest = normalize(dest);
+
+    float cosTheta = dot(start, dest);
```
const?

ilia-glushchenko commented on UnicornRender/utility/sources/Math.cpp:
```diff
@@ -0,0 +1,73 @@
+
+    rotationAxis = cross(start, dest);
+
+    float s = sqrt((1 + cosTheta) * 2);
```
const?

ilia-glushchenko commented on UnicornRender/utility/sources/Math.cpp:
```diff
@@ -0,0 +1,73 @@
+    rotationAxis = cross(start, dest);
+
+    float s = sqrt((1 + cosTheta) * 2);
+    float invs = 1 / s;
```
const?

ilia-glushchenko commented on UnicornRender/utility/sources/Math.cpp:
```diff
@@ -0,0 +1,73 @@
+glm::quat LookAt(glm::vec3 direction, glm::vec3 desiredUp, glm::vec3 worldY, glm::vec3 worldZ)
+{
+
+    if (length2(direction) < 0.0001f )
```
This threshold constant has to be declared explicitly in the header file.

ilia-glushchenko replied:
Please put the most likely case inside the `if`.

ilia-glushchenko commented on UnicornRender/utility/sources/Math.cpp:
```diff
@@ -0,0 +1,73 @@
+        return glm::quat();
+    }
+
+    glm::vec3 right = cross(direction, desiredUp);
```
const

ilia-glushchenko commented on UnicornRender/utility/sources/Math.cpp:
```diff
@@ -0,0 +1,73 @@
+    glm::vec3 right = cross(direction, desiredUp);
+    desiredUp = cross(right, direction);
+
+    glm::quat rot1 = RotationBetweenVectors(worldZ, direction);
```
const

ilia-glushchenko commented on UnicornRender/utility/sources/Math.cpp:
```diff
@@ -0,0 +1,73 @@
+    desiredUp = cross(right, direction);
+
+    glm::quat rot1 = RotationBetweenVectors(worldZ, direction);
+    glm::quat rot2 = RotationBetweenVectors(worldY, desiredUp);
```
Also could be moved to the return statement.

krupitskas replied:
This string will be to long and rot1 and rot2 have semantics. Its first and second rotation, and to multiply it correctly, multiplication must be reverse

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/CameraController.hpp:
```diff
@@ -0,0 +1,38 @@
+namespace video
+{
+
+/** @brief Camera controller class to create new controllers for cameras */
```
Can you change it so it would be more explicit that user code should not create instances of this class?

MrDmitry replied:
To be honest this class looks a bit redundant, it doesn't really add anything to `Transform` and user code always uses final child of this class. So this could just be removed completely and method/variable copy pasted in all needed controllers. Am I missing something?

krupitskas replied:
Yeah, at first I thought that this class will be much bigger and after of bunch development process it became legacy. Will delete it.

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/CameraFpsController.hpp:
```diff
@@ -4,38 +4,45 @@
+    UNICORN_EXPORT void UpdateView(float x, float y);
+
+    /** @brief Sets mouse coordinates without updating view matrix */
+    UNICORN_EXPORT void SetViewPositions(double x, double y);
```
Does this method really require double precision?

MrDmitry replied:
Or shall the other one also use double for consistency?

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/CameraFpsController.hpp:
```diff
@@ -4,38 +4,45 @@
-    bool m_dirty;
-    unicorn::video::Camera& m_camera;
+    /** @brief Calculates orientation in space */
+    virtual void CalculateOrientation() override;
```
model or world space?

krupitskas replied:
For now only world exist

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/Graphics.hpp:
```diff
@@ -143,9 +144,10 @@ class Graphics
 
     /** @brief  Spawn new Vulkan based Renderer, which contain VulkanDevice
     *  @param  window Output window.
+    *  @param  window Output window.
```
camera?

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/Mesh.hpp:
```diff
@@ -9,76 +9,94 @@
-    glm::mat4 modelMatrix;
+     * @brief Scales mesh
+     *
+     * @param[in] scale Each axis scale factor
```
`basis scale factors`?

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/Mesh.hpp:
```diff
@@ -9,76 +9,94 @@
+     *
+     * @param[in] scale Each axis scale factor
+     *
+     * {1,1,1} is origin scale, {0.5, 0.5, 2} is x and y diminished twice
```
`Lineary scales each basis component of the mesh vectors by the given scale factors.
Example:`

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/Mesh.hpp:
```diff
@@ -9,76 +9,94 @@
+    /**
+     * @brief Calculates model matrix for model view projection calculation
+     */
+    UNICORN_EXPORT void Calculate();
```
Maybe `CalculateModelMatrix` would be a better name than?

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/OrthographicCamera.hpp:
```diff
@@ -0,0 +1,47 @@
+public:
+    UNICORN_EXPORT OrthographicCamera(system::Window* window, glm::mat4& cameraProj);
+
+    UNICORN_EXPORT void UpdateProjection() override;
```
Comments.

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/OrthographicCamera.hpp:
```diff
@@ -0,0 +1,47 @@
+class OrthographicCamera final : public CameraProjection
+{
+public:
+    UNICORN_EXPORT OrthographicCamera(system::Window* window, glm::mat4& cameraProj);
```
Comments.

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/OrthographicCamera.hpp:
```diff
@@ -0,0 +1,47 @@
+    UNICORN_EXPORT void UpdateProjection() override;
+
+    /** @brief Changes scale by adding scaleOffset to it */
+    UNICORN_EXPORT void Scale(float scaleOffset);
```
Is this method really that necessary when you have getter and setter for the scale?
If it is `AddScale` or something with a verb would be a better name.

MrDmitry replied:
Yeah, I would remove this method when you have getter and setter. It just creates unnecessary contract.

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/OrthographicCamera.hpp:
```diff
@@ -0,0 +1,47 @@
+    /** @brief Changes scale by adding scaleOffset to it */
+    UNICORN_EXPORT void Scale(float scaleOffset);
+
+    /** @brief Sets scale which scaling ortho box */
```
`Sets orthogonal box scale factors`?

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/PerspectiveCamera.hpp:
```diff
@@ -0,0 +1,58 @@
+    */
+    UNICORN_EXPORT void Zoom(float zoomOffset);
+
+    /** @brief Sets fov */
```
`Sets field of view`?

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/PerspectiveCamera.hpp:
```diff
@@ -0,0 +1,58 @@
+    *
+    * @param zoomOffset adds it to fov value
+    */
+    UNICORN_EXPORT void Zoom(float zoomOffset);
```
`AddZoom`?

MrDmitry replied:
I'd just remove it

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,234 @@
+    UNICORN_EXPORT void Translate(glm::vec3 translate);
+
+    /**
+     * @brief Sets translate
```
`Sets translation vector`

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,234 @@
+namespace video
+{
+/**
+ * @brief Base class for every object that can translate or rotate
```
`that can be translated or rotated` passive voice

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,234 @@
+     *
+     * @param[in]  translate  The translate
+     */
+    UNICORN_EXPORT void Translate(glm::vec3 translate);
```
Is this a model or world space translation?

MrDmitry replied:
Wait is it the same thing as with FOV and Scale? If you have getter and setter there is no need for diff-based methods since they introduce additional (and probably unnecessary) interface.

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,234 @@
+    /**
+     * @brief Translates object
+     *
+     * @param[in]  translate  The translate
```
`translation vector`?

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,234 @@
+     *
+     * @param[in] direction new direction vector
+     */
+    UNICORN_EXPORT void LookAtDirection(glm::vec3 direction);
```
`SetLookAtDirection`

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,234 @@
+     * @param[in] direction new direction vector
+     * @param[in] upVector new up vector
+     */
+    UNICORN_EXPORT void LookAtDirection(glm::vec3 direction, glm::vec3 upVector);
```
`SetLookAtDirection`

MrDmitry replied:
Also the other `LookAtDirection`

MrDmitry replied:
Since it updates orientation maybe it makes more sense to call it `SetOrientation` and move it to existing `SetOrientation`?

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,234 @@
+     *
+     * @return direction
+     */
+    UNICORN_EXPORT glm::vec3 GetDirection() const;
```
`GetLookAtDirection`?

krupitskas replied:
This just a direction

MrDmitry replied:
direction of what? is it different from `LookAtDirection` method? If yes - describe how, if no - why the different naming?

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,234 @@
+     *
+     * @param[in]  translate new translate
+     */
+    UNICORN_EXPORT void SetTranslate(glm::vec3 translate);
```
Translate is a verb, translation is a noun, better use a noun instead of a verb.

MrDmitry replied:
And there already is a verb - Set

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,234 @@
+    virtual ~Transform() = default;
+
+    /**
+     * @brief Sets look at provided direction
```
`Sets the direction of the view?`

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,234 @@
+     *
+     * @return model matrix
+     *
+     * Don't forget to recalculate @sa Calculate
```
use `@note` or `@attention`?

MrDmitry replied:
Why should you recalculate after `GetModelMatrix`? I don't quite follow. Also move description before the `@return`

ilia-glushchenko commented on UnicornRender/video/source/CameraFpsController.cpp:
```diff
@@ -5,87 +5,59 @@
-
-    m_yaw += xoffset;
-    m_pitch += yoffset;
+    glm::quat x = glm::angleAxis(m_rotation.x, m_worldX);
```
const?

ilia-glushchenko commented on UnicornRender/video/source/CameraProjection.cpp:
```diff
@@ -0,0 +1,43 @@
+namespace video
+{
+
+CameraProjection::CameraProjection(system::Window* window, glm::mat4& cameraProj) :
```
If this constructor does not handles nullptr window, maybe take it by the reference?

ilia-glushchenko commented on UnicornRender/video/source/OrthographicCamera.cpp:
```diff
@@ -0,0 +1,59 @@
+    m_right = static_cast<float>(m_pWindow->GetSize().first) / 2;
+    m_top = static_cast<float>(m_pWindow->GetSize().second) / 2;
+
+    float xScaled = m_right * m_scaledAspect;
```
const

ilia-glushchenko commented on UnicornRender/utility/sources/Math.cpp:
```diff
@@ -0,0 +1,73 @@
+    if (cosTheta < -1 + 0.001f)
+    {
+        rotationAxis = cross(worldZ, start);
+        if (length2(rotationAxis) < 0.01 )
```
This threshold constant has to be declared explicitly in the header file.

ilia-glushchenko commented on UnicornRender/utility/sources/Math.cpp:
```diff
@@ -0,0 +1,73 @@
+        }
+        rotationAxis = normalize(rotationAxis);
+
+        return angleAxis(glm::radians(180.0f), rotationAxis);
```
remove redundant glm I guess.

ilia-glushchenko commented on UnicornRender/utility/sources/Math.cpp:
```diff
@@ -0,0 +1,73 @@
+    float s = sqrt((1 + cosTheta) * 2);
+    float invs = 1 / s;
+
+    return glm::quat(
```
I suggest you either don't use glm scope resolution or use it everywhere. 
In my opinion, second is better.

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/vulkan/Renderer.hpp:
```diff
@@ -57,8 +57,8 @@ struct SwapChainSupportDetails
 {
-    glm::mat4 view;
-    glm::mat4 proj;
+    glm::mat4 view = glm::mat4();
```
It does not look like this struct and 
```cpp
struct Camera
{
    glm::mat4 projection = glm::mat4();
    glm::mat4 view = glm::mat4();
}
```
has any difference as of now.

krupitskas replied:
Thats just a data for shader passing. But for consistent i will swap this two

dsodintsov replied:
Well, will you?

dsodintsov replied:
Also why is naming different? `proj` vs `projection`?

dsodintsov replied:
Order is still different from UnicornRender/video/include/unicorn/video/Camera.hpp

ilia-glushchenko commented on UnicornRender/video/source/Mesh.cpp:
```diff
@@ -43,5 +45,22 @@ const Material& Mesh::GetMaterial() const
+
+    auto T = glm::translate(glm::mat4(1.0), m_translation);
+    auto R = mat4_cast(m_orientation) * glm::mat4(1.0);
+    auto S = scale(glm::mat4(1.0), { m_scale });
```
I don't think that this is a good idea, at this point I can only make the difference between local and glm function calls just because we have a convention naming function starting with the capital letter. Half of the code uses `glm::` and the other half does not.

ilia-glushchenko commented on UnicornRender/utility/sources/Math.cpp:
```diff
@@ -0,0 +1,73 @@
+    float cosTheta = dot(start, dest);
+    glm::vec3 rotationAxis;
+
+    if (cosTheta < -1 + 0.001f)
```
Put the most likely case in the if statement, please.

MrDmitry replied:
And remove magic numbers

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/CameraProjection.hpp:
```diff
@@ -0,0 +1,54 @@
+    //! Aspect ratio
+    float m_aspect;
+    //! Reference to camera projection matrix
+    glm::mat4& m_cameraProjection;
```
All derived classes will not be copy constructible, is this really necessary?

MrDmitry replied:
I am not sure if copy/assign are needed for this class at all

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/Mesh.hpp:
```diff
@@ -9,76 +9,94 @@
+#include <glm/gtc/quaternion.hpp>
+#include <glm/gtc/matrix_transform.hpp>
 
 #include <list>
```
Not used anymore in this file.

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/PerspectiveCamera.hpp:
```diff
@@ -0,0 +1,58 @@
+    /** @brief Returns fov */
+    UNICORN_EXPORT float GetFov() const;
+
+    /** @brief      { function_description } */
```
is this supported by doxygen?

MrDmitry replied:
Is this memes?

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/vulkan/Renderer.hpp:
```diff
@@ -76,11 +76,12 @@ class Renderer : public video::Renderer
+     * @param[in] camera main camera
      */
-    Renderer(system::Manager& manager, system::Window* window);
+    Renderer(system::Manager& manager, system::Window* window, Camera& camera);
```
I don't see any information about what happens if you pass a nullptr as window parameter. Maybe better use a reference to force not null value, or mention the behavior in the documentation?

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/vulkan/Renderer.hpp:
```diff
@@ -76,11 +76,12 @@ class Renderer : public video::Renderer
 public:
     /**
-     * @brief Basic constructor
+     * @brief Constructor
```
`Constructs and initializes new renderer instance`?

ilia-glushchenko reviewed (changes requested):
Decide something on using `glm::` everywhere or not using it at all.

MrDmitry commented on UnicornRender/utility/include/unicorn/utility/Math.hpp:
```diff
@@ -0,0 +1,51 @@
+ *
+ * Realization : realtime rendering 3rd edition, 79 page
+ *
+ * @return returns a quaternion such that q*start = dest
```
omit `returns` in a `@return` please

ilia-glushchenko replied:
Not fixed.

MrDmitry commented on UnicornRender/utility/include/unicorn/utility/Math.hpp:
```diff
@@ -0,0 +1,51 @@
+ *
+ * @return orientation quaternion
+ */
+glm::quat LookAt(glm::vec3 direction, glm::vec3 desiredUp
```
If it's orientation, why is it called `LookAt`?

MrDmitry commented on UnicornRender/utility/include/unicorn/utility/Math.hpp:
```diff
@@ -0,0 +1,51 @@
+{
+
+/**
+ * @brief Computes rotation quaternion between two vectors
```
Calculates?

MrDmitry commented on UnicornRender/utility/include/unicorn/utility/Math.hpp:
```diff
@@ -0,0 +1,51 @@
+}
+}
+
+#endif // UNICORN_UTILITY_LOGGER_HPP
```
`LOGGER`?

MrDmitry commented on UnicornRender/video/include/unicorn/video/CameraFpsController.hpp:
```diff
@@ -4,38 +4,45 @@
+    /** @brief Sets mouse coordinates without updating view matrix */
+    UNICORN_EXPORT void SetViewPositions(double x, double y);
+
+    /** @brief Resets camera to provide setting new mouse coordinates */
```
to provide what?

MrDmitry commented on UnicornRender/video/include/unicorn/video/Material.hpp:
```diff
@@ -35,22 +34,35 @@ class Material
     UNICORN_EXPORT void SetIsWired(bool wireframe);
 
+    /**
+     * @brief Sets mesh in visible state
```
Mesh?

MrDmitry commented on UnicornRender/video/include/unicorn/video/Material.hpp:
```diff
@@ -35,22 +34,35 @@ class Material
+    UNICORN_EXPORT void SetIsVisible(bool visible);
+
+    /**
+     * @brief Determines if visible
```
How does it determine? If it just checks a boolean, that's just `Returns visibility flag`

MrDmitry commented on UnicornRender/video/include/unicorn/video/Material.hpp:
```diff
@@ -35,22 +34,35 @@ class Material
-    /**
-     * @brief Sets IsColored mode to true and removes binded texture
-     */
+    /** @brief Sets IsColored mode to true and removes binded texture */
```
bound, not binded

MrDmitry commented on UnicornRender/video/include/unicorn/video/Mesh.hpp:
```diff
@@ -9,76 +9,94 @@
     * @brief Sets new material
-    * @param [in] material updated material for current mesh
+    *
+    * @param[in] material updated material for current mesh
```
why is it `updated` when it's just `material to be used for current mesh`

MrDmitry commented on UnicornRender/video/include/unicorn/video/Renderer.hpp:
```diff
@@ -34,7 +34,13 @@ namespace video
-    Renderer(system::Manager& manager, system::Window* window);
+    /**
+     * @brief Constructor
+     * @param manager Describes required extensions
```
in/out?

MrDmitry replied:
will you fix this?

MrDmitry commented on UnicornRender/video/include/unicorn/video/Renderer.hpp:
```diff
@@ -34,7 +34,13 @@ namespace video
+    /**
+     * @brief Constructor
+     * @param manager Describes required extensions
+     * @param window Render into
```
render into? also in/out?

MrDmitry replied:
will you fix this?

MrDmitry commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,234 @@
+    UNICORN_EXPORT glm::vec3 GetDirection() const;
+
+    /**
+     * @brief Return right vector
```
What is `right vector`? Is it just axis like `forward-right-up`?

krupitskas replied:
Yes, this is world vectors multiplied by camera orientation.

MrDmitry commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,234 @@
+     *
+     * @param[in] distance distance to translate
+     */
+    UNICORN_EXPORT void TranslateLocalX(float distance);
```
What is the point of this (and similar) methods when you have getter and setter? User code can decide how it wants to modify it

krupitskas replied:
This is part of engine, its tool. Local translates requires orientation premultiplication. Thats why this methods exist. If user will get translate and add by X new one, this will be world X translation

MrDmitry commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,234 @@
+     *
+     * @param[in] rotation amount of radians at each axis
+     */
+    UNICORN_EXPORT void Rotate(glm::vec3 rotation);
```
I'd just give getter and setter instead

MrDmitry commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,234 @@
+     * @param[in] axis axis of rotation
+     * @param[in] point point of interest
+     */
+    UNICORN_EXPORT void RotateAroundPoint(float radians, glm::vec3 axis, glm::vec3 point);
```
This feels like utility method that usercode shall use on its own, not via class interface

MrDmitry commented on UnicornRender/video/include/unicorn/video/vulkan/Renderer.hpp:
```diff
@@ -76,11 +76,12 @@ class Renderer : public video::Renderer
-     * @brief Basic constructor
+     * @brief Constructor
      * @param manager Describes required extensions
      * @param window Render into
```
render info?

MrDmitry commented on UnicornRender/video/include/unicorn/video/vulkan/VkMesh.hpp:
```diff
@@ -89,7 +89,7 @@ class VkMesh
     /** @brief Returns @c true if mesh is in wired mode and @c false otherwise */
     bool IsWired() const;
-
+    bool isVisible() const;
```
doxy

MrDmitry commented on UnicornRender/video/source/Camera2DController.cpp:
```diff
@@ -0,0 +1,24 @@
+
+Camera2DController::Camera2DController(glm::mat4& cameraView) :
+    CameraController(cameraView)
+{
```
looks like a dummy class

MrDmitry commented on UnicornRender/video/source/CameraFpsController.cpp:
```diff
@@ -5,87 +5,59 @@
 }
 
-void CameraFpsController::UpdateView(double posX, double posY)
+void CameraFpsController::CalculateOrientation()
```
if it doesn't return the calculated value (and I'm not saying it should), you could name it `UpdateOrientation` instead since that's what it does - it updates orientation and rotation

MrDmitry commented on UnicornRender/video/source/OrthographicCamera.cpp:
```diff
@@ -0,0 +1,59 @@
+    , m_scaledAspect(m_aspect / m_orthoScale)
+    , m_right(static_cast<float>(window->GetSize().first) / 2)
+    , m_top(static_cast<float>(m_pWindow->GetSize().second) / 2)
+    , m_back(-1000.f)
```
magic number?

krupitskas replied:
Just initial size of box, 1km

MrDmitry commented on UnicornRender/video/source/PerspectiveCamera.cpp:
```diff
@@ -0,0 +1,64 @@
+    , m_fovLowerBound(44.f)
+    , m_fovUpperBound(45.f)
+    , m_znear(0.1f)
+    , m_zfar(100000.f) // 100 kilometers
```
magic numbers

krupitskas replied:
Why is it?

MrDmitry replied:
Why 100000? if there is no reason, this is a magic number that you just decided to use. Move it to header file and make it a constant there.

MrDmitry commented on UnicornRender/video/source/Transform.cpp:
```diff
@@ -0,0 +1,204 @@
+    m_isDirty = true;
+}
+
+void Transform::Rotate(float angleRadians, glm::vec3 axis) {
```
`{`

MrDmitry commented on UnicornRender/video/source/vulkan/Renderer.cpp:
```diff
@@ -1309,6 +1304,12 @@ bool Renderer::CreateCommandBuffers()
             {
+                if (!pVkMesh->isVisible())
+                {
+                    ++j;
```
Could you briefly explain why `j` shall be incremented here? It's used in `dynamicOffset` afterwards but if a mesh is invisible, is it still stored somewhere in a pipeline or how does this work?

krupitskas replied:
++j said that we just skip this mesh and move to next one. We just not rendering it.

Commit:
[#86] Fixed 1.65 compability, some review fixes

Commit:
[#86] unnecessary function delettion

Commit:
[#86] Lots of refactoring and doxy fixing

Commit:
[#86] Removed assimp folder

ilia-glushchenko commented on UnicornRender/utility/sources/Math.cpp:
```diff
@@ -53,21 +55,30 @@ glm::quat RotationBetweenVectors(glm::vec3 start, glm::vec3 dest, glm::vec3 worl
 {
-
-    if (length2(direction) < 0.0001f )
+    if (length2(direction) < 0.0001f) // Already look at desired direction
```
Global floating point comparison to the threshold.

ilia-glushchenko commented on UnicornRender/utility/sources/Math.cpp:
```diff
@@ -53,21 +55,30 @@ glm::quat RotationBetweenVectors(glm::vec3 start, glm::vec3 dest, glm::vec3 worl
+{
+    glm::vec3 outputTranslation;
+    glm::vec3 dir = point - originalTranslation;
+    outputTranslation = originalTranslation + dir;
```
The definition can be merged with the declaration.

krupitskas replied:
Resharper boii

ilia-glushchenko commented on UnicornRender/utility/include/unicorn/utility/Math.hpp:
```diff
@@ -0,0 +1,67 @@
+ * @param[in] worldX world X axis
+ * @param[in] worldZ world Z axis
+ *
+ * Realisation : Realtime rendering 3rd edition, 79 page by Eric Haines, Naty Hoffman, and Tomas M�ller
```
Implementation

ilia-glushchenko commented on UnicornRender/utility/sources/Math.cpp:
```diff
@@ -0,0 +1,84 @@
+    float const cosTheta = glm::dot(start, dest);
+    glm::vec3 rotationAxis;
+
+    float const floatError = 0.001f;
```
This constant should be in the unified head file so that the whole system would have the same degree of accuracy.

ilia-glushchenko commented on UnicornRender/utility/sources/Math.cpp:
```diff
@@ -0,0 +1,84 @@
+
+    float const floatError = 0.001f;
+
+    if (cosTheta < -1 + floatError) // they look in different sides
```
Put the most likely case in the if statement.

ilia-glushchenko commented on UnicornRender/utility/sources/Math.cpp:
```diff
@@ -0,0 +1,84 @@
+
+glm::quat LookAt(glm::vec3 direction, glm::vec3 desiredUp, glm::vec3 worldY, glm::vec3 worldZ)
+{
+    if (length2(direction) < 0.0001f) // Already look at desired direction
```
I think that you can neglect this check, And simply specify it in the documentation. 
Also, I think that this should be called `CalculateOrientationQuaternion` because it gives a more precise mathematical meaning behind the function. And if you desire to have a convention name like `LookAt` you can simply declare an alias or an inline proxy function.

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/Camera2DController.hpp:
```diff
@@ -0,0 +1,37 @@
+     */
+    UNICORN_EXPORT Camera2DController(glm::mat4& cameraView);
+
+    UNICORN_EXPORT void Update();
```
Documentation.

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/CameraFpsController.hpp:
```diff
@@ -4,38 +4,50 @@
+     * in next UpdateView call */
+    UNICORN_EXPORT void ResetView();
+
+    UNICORN_EXPORT void Update();
```
Documentation.

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/CameraFpsController.hpp:
```diff
@@ -4,38 +4,50 @@
+    glm::vec2 m_mousePosition;
+    bool m_dirtyViewPosition;
+
+    glm::mat4& m_cameraView;
```
What's the logic behind this class not being copy constructible? It does not look like it manages any resources.

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/Mesh.hpp:
```diff
@@ -9,76 +9,91 @@
     UNICORN_EXPORT Mesh(Material const& material);
+
+    /** @brief Default destructor */
     UNICORN_EXPORT ~Mesh() = default;
```
What is the need of a default public destructor?

ilia-glushchenko commented on UnicornRender/video/source/Transform.cpp:
```diff
@@ -0,0 +1,186 @@
+
+void Transform::UpdateOrientation()
+{
+    glm::quat z = glm::angleAxis(m_rotation.z, m_worldZ);
```
All of it can be `const`

Commit:
[#86] Index buffer type fix

Commit:
[#86] Transform calss modification, some review small changes

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -106,46 +106,28 @@ class Transform
 
     /**
-     * @brief Translates by local Y
+     * @brief Translates by axises
```
`axes`

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -106,46 +106,28 @@ class Transform
      *
-     * @param[in] distance distance to translate
+     * Lineary scales each basis component of the mesh vectors by the given scale factors
+     * {1,1,1} is origin scale, {0.5, 0.5, 2} is x and y diminished twice
```
`Example:...`?

ilia-glushchenko commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -106,46 +106,28 @@ class Transform
+    UNICORN_EXPORT void TranslateLocal(glm::vec3 distance);
     /**
-     * @brief Translates by local Z
+     * @brief Translates by world axises
```
`axes`

MrDmitry commented on UnicornRender/video/include/unicorn/video/Graphics.hpp:
```diff
@@ -142,10 +143,11 @@ class Graphics
     UNICORN_EXPORT void SetWindowCreationHint(system::WindowHint hint, int32_t value) const;
 
     /** @brief  Spawn new Vulkan based Renderer, which contain VulkanDevice
```
Vulkan based or DriverType based?

MrDmitry replied:
will you fix this?

MrDmitry commented on UnicornRender/video/include/unicorn/video/Mesh.hpp:
```diff
@@ -9,77 +9,73 @@
+    /**
+     * @brief Constructs mesh
+     *
+     * @param[in] material visual describing material
```
visual appearance description?

MrDmitry commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,214 @@
+     * @param[in]  y world axis
+     * @param[in]  z world axis
+     */
+    UNICORN_EXPORT void SetWorldCoordinates(glm::vec3 x, glm::vec3 y, glm::vec3 z);
```
This doesn't set world coordinates, it sets world axes. Rename?

MrDmitry commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,214 @@
+     *
+     * @param[in] radians amount of radians
+     */
+    UNICORN_EXPORT void RotateX(float radians);
```
I think `RotateX/Y/Z` can be removed since you have `void Rotate(float angleRadians, glm::vec3 axis);` or you can inline them in the header to proxy the call to `Rotate(angle, axis)`

MrDmitry commented on UnicornRender/video/source/OrthographicCamera.cpp:
```diff
@@ -0,0 +1,54 @@
+    , m_right(static_cast<float>(m_pWindow->GetSize().first) / 2)
+    , m_top(static_cast<float>(m_pWindow->GetSize().second) / 2)
+    , m_back(-1000.f)
+    , m_front(1000.f)
```
please move constants to header file to avoid magic numbers

dsodintsov commented on UnicornRender/video/include/unicorn/video/CameraFpsController.hpp:
```diff
@@ -4,38 +4,51 @@
-    bool m_dirty;
-    unicorn::video::Camera& m_camera;
+    /** @brief Calculates orientation in space */
+    virtual void UpdateOrientation() override;
```
Why is it in `private` section? It's `public` in `Transform`

Commit:
[#86] Mistakes fixed, redundant rotateX removed

[#86] Camera const ref fix, documentation fix

[#86] Transform documentation refactoring

[#86] More documentation work

[#86] controllers fix

ilia-glushchenko commented on UnicornRender/utility/sources/Math.cpp:
```diff
@@ -0,0 +1,82 @@
+
+glm::quat CalculateOrientationQuaternion(glm::vec3 direction, glm::vec3 desiredUp, glm::vec3 worldY, glm::vec3 worldZ)
+{
+    if (length2(direction) < SystemAccuracy()) // Already look at desired direction
```
glm

ilia-glushchenko commented on UnicornRender/utility/sources/Math.cpp:
```diff
@@ -0,0 +1,82 @@
+        return glm::quat();
+    }
+
+    glm::vec3 const right = cross(direction, desiredUp);
```
glm

MrDmitry replied:
Please do. Also on the next line

ilia-glushchenko commented on UnicornRender/video/source/CameraFpsController.cpp:
```diff
@@ -5,87 +5,64 @@
-    m_lastX = posX;
-    m_lastY = posY;
+    glm::quat x = glm::angleAxis(m_rotation.x, m_worldX);
+    glm::quat y = glm::angleAxis(m_rotation.y, conjugate(m_orientation) * m_worldY);
```
glm

MrDmitry commented on UnicornRender/video/source/PerspectiveCamera.cpp:
```diff
@@ -18,8 +18,8 @@ PerspectiveCamera::PerspectiveCamera(system::Window& window, glm::mat4& cameraPr
-    , m_znear(0.1f)
-    , m_zfar(100000.f) // 100 kilometers
+    , m_znear(NEAR_PERSPECTIVE_FRUSTRUM_PLANE)
+    , m_zfar(FAR_PERSPECTIVE_FRUSTRUM_PLANE) // 100 kilometers
```
I think comment no longer makes sense

MrDmitry commented on UnicornRender/video/include/unicorn/video/CameraFpsController.hpp:
```diff
@@ -4,38 +4,51 @@
+    /** @brief Sets mouse coordinates without updating view matrix */
+    UNICORN_EXPORT void SetViewPositions(float x, float y);
+
+    /** @brief Sets flag to update mouse position without recalculation of orientation
```
that's not a brief anymore, split the description

MrDmitry commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,193 @@
+    UNICORN_EXPORT void LookAtDirection(glm::vec3 direction, glm::vec3 upVector);
+
+    /**
+     * @brief Sets up vector
```
I believe it doesn't just set `up vector` but it updates orientation instead. Please make sure that your description matches visible side effects. In this case calling `SetUp` and eventual `UpdateModelMatrix` would change all vectors so it updates more than just `up vector`

MrDmitry commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,193 @@
+    UNICORN_EXPORT void SetUp(glm::vec3 upVector);
+
+    /**
+     * @brief      Sets the world coordinates
```
coordinates?

MrDmitry commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,193 @@
+    UNICORN_EXPORT void SetWorldAxes(glm::vec3 x, glm::vec3 y, glm::vec3 z);
+
+    /**
+     * @brief Sets translate
```
that's not very descriptive. What does it mean?

MrDmitry commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,193 @@
+    UNICORN_EXPORT glm::vec3 GetUp() const;
+
+    /**
+     * @brief Return translate
```
That's not very descriptive. Translate from what?

MrDmitry commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,193 @@
+    /**
+     * @brief Returns model matrix
+     *
+     * @attention Don't forget to recalculate @sa Calculate before
```
That's not a good place for this `@attention` - at the point when you want to `GetModelMatrix()` you do not know if it was modified. And suggesting calling `Calculate()` each time when you want to get the model matrix does not make much sense. I would suggest moving this attention to all methods that may lead to model matrix modifications:
`Don't forget to call Calculate() to commit all changes to the object`

MrDmitry replied:
Also as a side note - I don't see `Calculate()` method here

MrDmitry commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,193 @@
+     *
+     * @param[in] scale basis scale factors
+     *
+     * Lineary scales each basis component of the mesh vectors by the given scale factors
```
`Linearly`

Also move the description _before_ `@param`
Check https://github.com/Godlike/Unicorn/wiki/Coding-style-guideline#class-comments

MrDmitry commented on UnicornRender/video/include/unicorn/video/vulkan/VkMesh.hpp:
```diff
@@ -90,6 +90,9 @@ class VkMesh
     /** @brief Returns @c true if mesh is in wired mode and @c false otherwise */
     bool IsWired() const;
 
+    /** @brief Returns true if mesh visible, false otherwise */
```
`if mesh is visible`

MrDmitry commented on UnicornRender/video/include/unicorn/video/vulkan/VkMesh.hpp:
```diff
@@ -90,6 +90,9 @@ class VkMesh
     bool IsWired() const;
 
+    /** @brief Returns true if mesh visible, false otherwise */
+    bool isVisible() const;
```
UpperCamelCase

MrDmitry commented on UnicornRender/video/include/unicorn/video/vulkan/VkMesh.hpp:
```diff
@@ -102,7 +105,7 @@ class VkMesh
 
     /**
-     * @brief Material in vulkan is combination of descriptor set and binded data   
+     * @brief Material in vulkan is combination of descriptor set and binded data
```
`is a combination`
`bound data`

MrDmitry commented on UnicornRender/video/include/unicorn/video/vulkan/Renderer.hpp:
```diff
@@ -76,11 +76,12 @@ class Renderer : public video::Renderer
+     * @param[in] camera main camera
      */
-    Renderer(system::Manager& manager, system::Window* window);
+    Renderer(system::Manager& manager, system::Window* window, Camera& camera);
```
why aren't manager and camera const references?

MrDmitry reviewed (changes requested):
There are unanswered comments from Ilya (and a couple of mine) I tried to mark those with a question or a 👍

MrDmitry commented on UnicornRender/video/include/unicorn/video/Mesh.hpp:
```diff
@@ -9,77 +9,73 @@
+    /**
+     * @brief Constructs mesh
+     *
+     * @param[in] material visual appearance description
```
I would suggest pass by value since we're copying the material anyway

dsodintsov commented on UnicornRender/video/include/unicorn/video/Mesh.hpp:
```diff
@@ -9,77 +9,73 @@
     * @brief Sets new material
-    * @param [in] material updated material for current mesh
+    *
+    * @param[in] material material to be used for current mesh
```
I would suggest pass by value since we're copying the material anyway

MrDmitry commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,202 @@
+     */
+    UNICORN_EXPORT void SetRotation(glm::vec3 rotation);
+
+    /** @brief Recalculates model matrix */
```
There is no `model matrix` in the class. This method updates several vectors and calculates `m_transformMatrix`

Please update documentation. Also method name does not make much sense - UpdateTransformMatrix?

MrDmitry commented on UnicornRender/video/include/unicorn/video/Transform.hpp:
```diff
@@ -0,0 +1,202 @@
+    /**
+     * @brief Returns direction vector
+     *
+     * @attention Don't forget to recalculate @sa UpdateModelMatrix to commit all changes to the object
```
I suggest moving this `@attention` to setters and other methods that invalidate the state of the object, not getters - it makes more sense to mention this in setters since that's when recalculation/invalidation has to happen.

MrDmitry commented on UnicornRender/video/include/unicorn/video/Graphics.hpp:
```diff
@@ -141,11 +142,12 @@ class Graphics
 
-    /** @brief  Spawn new Vulkan based Renderer, which contain VulkanDevice
-    *  @param  window Output window.
+    /** @brief  Spawn new DriverType based Renderer, which contain VulkanDevice
```
VulkanDevice?

ilia-glushchenko commented:
Approve

Commit:
[#86] Changing CI vulkan version and location

ilia-glushchenko commented:
Approve
